[SemVer] adding grammar for the semantic version spec

myarichuk opened the pull request:
This is a grammar for [Semantic Version](https://semver.org/) spec with a small addition (that doesn't violate the spec but allows for a more straightforward interpretation of the parse results).
I've added tokens that would recognize "build tags" such as "alpha3" or "beta.4" -> this should allow developers easier implementation of SemVer comparison.

myarichuk commented:
Note: the work on the parser is not yet finished so it is likely I will submit another PR for the commit with both C# "interpreter" that uses the grammar and relevant unit tests, which can also serve as examples of the semver spec and it's edge-cases.  Also, since I am a noob with ANTLR, I'd love to see any feedback :)
  
* WIP of a project that uses the grammar can be seen [here](https://github.com/myarichuk/ReleaseTools/tree/master/src/Parser.ConventionalCommit)
* Unit test project (also WIP) can be seen [here](https://github.com/myarichuk/ReleaseTools/blob/master/tests/ParserTests/SemanticVersion.cs)

kaby76 commented:
> Note: the work on the parser is not yet finished so it is likely I will submit another PR for the commit with both C# "interpreter" that uses the grammar (example of intended usage of the grammar) and relevant unit tests, which can serve as examples of the semver spec and it's edge-cases. Also, since I am a noob with ANTLR, I'd love to see any feedback :)
> 
> * WIP of a project that uses the grammar can be seen [here](https://github.com/myarichuk/ReleaseTools/tree/master/src/Parser.ConventionalCommit)
> * Unit test project (also WIP) can be seen [here](https://github.com/myarichuk/ReleaseTools/blob/master/tests/ParserTests/SemanticVersion.cs)

Do not include driver code for your grammar. The grammar is tested by generating a driver from template with the Github Actions build. https://github.com/antlr/grammars-v4/tree/master/_scripts/templates

Include a pom.xml in the directory "grammars-v4/semver/". Make a copy from something like grammars-v4/abb/pom.xml. 

Include a line in the "root pom.xml" a reference to your new grammar, somewhere around [here](https://github.com/antlr/grammars-v4/blob/87bcdac9d4487e73ac6566d07e44161915c2bb85/pom.xml#L264) (alphabetic sorting).

Include a directory of test files in "grammars-v4/semver/examples/". All files in that are considered source except ".errors" and ".tree" files.

myarichuk commented:
Awesome, thanks for so many detailed suggestions! Plan to follow the suggestions as soon as I have enough free time.

KvanTTT commented on semver/SemanticVersion.g4:
```diff
@@ -0,0 +1,98 @@
+
+versionCore: major DOT minor DOT patch;
+
+semver:
+    versionCore                               #SemverWithCore
+  | versionCore DASH preRelease               #SemverWithPreRelease
+  | versionCore PLUS build                    #SemverWithBuild
+  | versionCore DASH preRelease PLUS build    #SemverWithPreReleaseAndBuild
+  ;
```
Honeslty, I'm not sure such splitting on alternaitves makes big sense. I suggest simplifying:

```suggestion
semver
  : versionCore (DASH preRelease (PLUS build)? | PLUS build)?
  ;
```

KvanTTT reviewed (changes requested):
Welcome to the grammars repository and thanks for the contribution.

But some issues should be fixed.

kaby76 commented on semver/SemanticVersion.g4:
```diff
@@ -0,0 +1,98 @@
+grammar SemanticVersion;
+
+fragment A : [aA];
+fragment B : [bB];
+fragment C : [cC];
+fragment D : [dD];
+fragment E : [eE];
+fragment F : [fF];
+fragment G : [gG];
+fragment H : [hH];
+fragment I : [iI];
+fragment J : [jJ];
+fragment K : [kK];
+fragment L : [lL];
+fragment M : [mM];
+fragment N : [nN];
+fragment O : [oO];
+fragment P : [pP];
+fragment Q : [qQ];
+fragment R : [rR];
+fragment S : [sS];
+fragment T : [tT];
+fragment U : [uU];
+fragment V : [vV];
+fragment W : [wW];
+fragment X : [xX];
+fragment Y : [yY];
+fragment Z : [zZ];
```
If caseInsensitive is used, it'll no longer work with Antlr4cs. I wasn't sure if it was required.

myarichuk replied:
Ideally, I would like the grammar to be as compatible as possible. Wasn't aware of "caseInsensitive" feature, I will definitely use it for my own stuff :)
@KvanTTT - thanks for the detailed review and suggestions. They make perfect sense, I will address them.

KvanTTT replied:
> it'll no longer work with Antlr4cs.

I don't consider it's useful to support an outdated runtime.

KvanTTT replied:
Also, we use `caseInsensitive` option for other grammars in this repository.

myarichuk commented:
Addressed the review suggestions, thanks again for your detailed feedback!

* Added a "local" ``pom.xml`` in the grammar folder
* Split the combined grammar into lexer/parser files
* Used the "caseIncensitive" flag for the lexer.
* Removed unnecessary parentheses from lexer token definitions
* Removed the alternative splitting for the ``semver`` rule - it makes more sense to combine them, yeah.
* Updated the "global" ``pom.xml`` to include the grammar name. I assume the name there should have the same name mentioned in the ``pom.xml`` in the grammar folder, right?
* Added some examples in the ``examples`` folder - one example per text file. Judging by what I see in other folder that should work, right?

I wonder if it makes sense to include a version with case-insensitive fragments for people who use the **Antlr4cs**. Wdyt?

myarichuk commented:
note: missed ``artifactId`` from pom.xml - looks like it made the build fail. I changed ``artifactId`` so it would be unique.

KvanTTT commented on semver/README.md:
````diff
@@ -0,0 +1,58 @@
+fragment B : [bB];
+fragment C : [cC];
+fragment D : [dD];
+fragment E : [eE];
+fragment F : [fF];
+fragment G : [gG];
+fragment H : [hH];
+fragment I : [iI];
+fragment J : [jJ];
+fragment K : [kK];
+fragment L : [lL];
+fragment M : [mM];
+fragment N : [nN];
+fragment O : [oO];
+fragment P : [pP];
+fragment Q : [qQ];
+fragment R : [rR];
+fragment S : [sS];
+fragment T : [tT];
+fragment U : [uU];
+fragment V : [vV];
+fragment W : [wW];
+fragment X : [xX];
+fragment Y : [yY];
+fragment Z : [zZ];
+
+fragment ALPHA: A L P H A;
+fragment BETA: B E T A;
+fragment RC: (R C) | (R E L E A S E (DASH | DOT) C A N D I D A T E);
+fragment SNAPSHOT: S N A P S H O T;
+fragment PREVIEW: P | P R E | P R E V I E W;
+fragment DEV: D E V | D E V E L | D E V E L O P M E N T;
+fragment MILESTONE: M T | M I L E S T O N E;
+fragment DAILY: D A I L Y;
+fragment NIGHTLY: N I G H T L Y;
+fragment BUILD: B L D | B U I L D;
+fragment TEST: T E S T;
+fragment EXPERIMENTAL: E X P E R I M E N T A L;
+
+```
````
Sorry, but I don't consider this info useful, we don't have such info in other case insensitive grammars (check [`sql`](https://github.com/antlr/grammars-v4/tree/master/sql) for example). It's well-known practice to use case insensitive fragments. Also, I think there is a small number of outdated Antlr4cs runtime.

```suggestion
```

myarichuk replied:
Makes sense, I will remove this section in the readme

KvanTTT commented on semver/SemanticVersionParser.g4:
```diff
@@ -0,0 +1,27 @@
+build:
+      value = NUMBER                                             #BuildNumber
+    | DASH                                                       #BuildDash
+    | tag = TAG (DOT | DASH)? version = (NUMBER | IDENTIFIER)?   #BuildTagged
```
If `tag` is parser rule, then the following is actual:

```suggestion
    | tag (DOT | DASH)? version = (NUMBER | IDENTIFIER)?   #BuildTagged
```

KvanTTT commented on semver/SemanticVersionParser.g4:
```diff
@@ -0,0 +1,27 @@
+preRelease:
+      value = NUMBER                                            #PreReleaseNumber
+    | DASH                                                      #PreReleaseDash
+    | tag = TAG (DOT | DASH)? version = (NUMBER | IDENTIFIER)?  #PreReleaseTagged
```
```suggestion
    | tag (DOT | DASH)? version = (NUMBER | IDENTIFIER)?  #PreReleaseTagged
```

KvanTTT commented on semver/SemanticVersionParser.g4:
```diff
@@ -0,0 +1,27 @@
+    | left = preRelease (DOT | DASH)? right = preRelease        #PreReleaseIdentifierExtended
+    ;
+
+versionCore: major DOT minor DOT patch;
```
The following definition is shorter:

```suggestion
versionCore: major=NUMBER DOT minor=NUMBER DOT patch=NUMBER;
```

KvanTTT commented on semver/SemanticVersionParser.g4:
```diff
@@ -0,0 +1,27 @@
+
+options { tokenVocab = SemanticVersionLexer; }
+
+patch: value = NUMBER;
+minor: value = NUMBER;
+major: value = NUMBER;
```
No need in these rules if use [rule element labels](https://github.com/antlr/antlr4/blob/master/doc/parser-rules.md#rule-element-labels) (see below).

```suggestion
```

KvanTTT reviewed (changes requested):
Thanks, but some issues are still presented.

myarichuk commented:
Hey, did some fixes. The test show pass now (``grammarName`` in the ``antlr4test`` config was wrong so the plugin couldn't find the generated grammar classes)
Also addressed the comments.

Thanks for the detailed comments!

teverett commented:
@KvanTTT  good to merge this?

KvanTTT reviewed (approved):
Thanks again for the new grammar and its fixes!

myarichuk commented:
Thanks for the detailed feedback, I really appreciate it!
